backend/books/serializers.py

```diff
@@ -130,17 +130,43 @@ def get_line_items(self, instance):
             sale['id'] = sale.pop('sales_reconciliation')
 
         buybacks = Buyback.objects.filter(book=instance.id).annotate(date=F('buyback_order__date')).annotate(vendor=F('buyback_order__vendor')).annotate(
-            vendor_name=F('buyback_order__vendor__name')).annotate(unit_price=F('unit_buyback_price')).annotate(type=Value("buyback order", CharField())).values(
-                'buyback_order', 'date', 'vendor', 'vendor_name', 'unit_price', 'quantity', 'type')
+            vendor_name=F('buyback_order__vendor__name')).annotate(unit_price=F('unit_buyback_price')).annotate(username=F('buyback_order__user__username')).annotate(type=Value("buyback order", CharField())).values(
+                'buyback_order', 'date', 'vendor', 'vendor_name', 'unit_price', 'quantity', 'type', 'username')
 
         buybacks = list(buybacks)
         for buyback in buybacks:
             buyback['id'] = buyback.pop('buyback_order')
+        
+        corrections = BookInventoryCorrection.objects.filter(book=instance.id).annotate(quantity=F('adjustment')).annotate(
+            username=F('user__username')).annotate(type=Value("inventory corrections", CharField())).values('id', 'date', 'quantity', 'type', 'username')
+
+        corrections = list(corrections)
+
+        items = purchases + sales + buybacks + corrections
 
-        items = purchases + sales + buybacks
-        for item in items:
+        # Sort it by date
+        sorted_items = sorted(items, key=lambda d: d['date'])
+
+        # Add running stock
+        stock = 0
+        for item in sorted_items:
+            # change the date str format
             item['date'] = item['date'].strftime('%Y-%m-%d')
-        return items
+            change_in_stock = self.get_delta_stock(item)
+            stock += change_in_stock
+            item['stock'] = stock
+
+        return sorted_items
```

Comment:
Not sure I am a fan of this approach. If my understanding is correct, you are sorting all transactions with the given book, with earliest transaction first, where the inventory count can be assumed to be zero, and then calculating from the beginning iteratively each inventory count? This would mean that at the end, the `stock` variable should equal the current inventory? I'm just wondering if this approach could ever run into issues. Would prefer if we could store this value... but also calculation speed isn't a concern we should consider atm. So actually I think this should be okay. I do think it can't hurt to put a check to ensure the ending calculated `stock` does indeed equal the current inventory, as a sanity check and for readability to help code make it clear what is being done here.

Reply:
added assert statement